pkg/core/tekton/tekton.go

```diff
@@ -0,0 +1,403 @@
+	if err != nil {
+		if !k8serr.IsNotFound(err) {
+			return "", err
+		}
+		logger.Printf("Creating tekton trigger binding for workflow: %s...", workflowName)
+		_, err := t.tektonClients.TriggerBindingClient.Create(ctx, triggerBinding, metav1.CreateOptions{})
+		if err != nil {
+			return "", err
+		}
+	}
+
+	eventListener := generateEventListener(triggerTemplate, triggerBinding)
+	el, err := t.tektonClients.EventListenerClient.Get(ctx, workflowName, metav1.GetOptions{})
+	if err != nil {
+		if !k8serr.IsNotFound(err) {
+			return "", err
+		}
+		logger.Printf("Creating tekton event listener for workflow: %s...", workflowName)
+		el, err = t.tektonClients.EventListenerClient.Create(ctx, eventListener, metav1.CreateOptions{})
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if err := waitFor(t.eventListenerReady(ctx, el.Name), 1*time.Second, 1*time.Minute); err != nil {
+		return "", fmt.Errorf("event listener %q did not get ready in the expected time", el.Name)
+	}
+
+	el, _ = t.tektonClients.EventListenerClient.Get(ctx, workflowName, metav1.GetOptions{})
+	return el.Status.Address.URL.String(), nil
+}
+
+func waitFor(waitFunc wait.ConditionFunc, interval, timeout time.Duration) error {
+	return wait.PollImmediate(interval, timeout, waitFunc)
+}
+
+func (t *WorkflowBackend) eventListenerReady(ctx context.Context, name string) wait.ConditionFunc {
+	return func() (bool, error) {
+		el, err := t.tektonClients.EventListenerClient.Get(ctx, name, metav1.GetOptions{})
+		if err != nil {
+			return false, nil
+		}
+		// No conditions have been set yet
+		if len(el.Status.Conditions) == 0 {
+			return false, nil
+		}
+		if el.Status.GetCondition(apis.ConditionType(appsv1.DeploymentAvailable)) == nil {
+			return false, nil
+		}
+		for _, cond := range el.Status.Conditions {
+			if cond.Status != corev1.ConditionTrue {
+				return false, nil
+			}
+		}
+		if el.Status.Address.URL == nil {
+			return false, nil
+		}
+		return true, nil
+	}
+}
+
+func generatePipeline(w workflow.Workflow, namespace string) *v1beta1.Pipeline {
+	resolver := newVariablesResolver()
+	pb := builder.NewPipelineBuilder(w.Name, namespace)
+	pb.PipelineMeta(builder.Label("fuseml/generated-from", w.Name))
+	pb.PipelineDescription(*w.Description)
+
+	// process the FuseML workflow inputs
+	for _, input := range w.Inputs {
+		// an input of the 'codeset' type means a git repository for tekton.
+		// adds a workspace, resource and the clone task to the pipeline, also
+		// also add 'codeset-name' and 'codeset-version' parameters to the tekton
+		// pipeline which represents codeset.name and codeset.version in a FuseML
+		// workflow.
+		if *input.Type == inputTypeCodeset {
+			pb.PipelineWorkspace(codesetWorkspaceName)
+			pb.PipelineResource("source-repo", "git")
+			pb.PipelineTask("clone", cloneTaskName, nil, map[string]string{codesetWorkspaceName: codesetWorkspaceName},
+				map[string]string{"source-repo": "source-repo"})
```

Comment:
Why are the keys in these maps (workspaces/resources) same as values?

Reply:
What is happening is:
1. At the pipeline level we declare a workspace with name `codesetWorkspaceName` and a resource with name `source-repo`
2. When calling the task we need to pass the workspace and resource to it, they have the same name because the task also uses this same name to reference the workspace and the resource,